### PR TITLE
Refactored DataObjectSerializer to be able to use decimal value types and BsonRepresentation attributes

### PR DESCRIPTION
**Pull request by MrZoidberg**
See #761. Previous DataObjectSerializer class has many problems listed in the issue. I've fixed this problem by introducing the ability to use class maps for data classes and enriching serialized BSON with type information by using reflection. This is purely optional and old method is used by default.

### CODE REVIEW

**Review by danielgerlag — APPROVED** [review]
This is great! 
Can you add a section to the MongoDb readme explaining this?
Also, could you increment the version number of the package, thanks!

**Comment by MrZoidberg** [comment]
@danielgerlag sure, will do in the nearest days when I have some free time

**Comment by MrZoidberg** [comment]
@danielgerlag done. Please check.

File: src/WorkflowCore/WorkflowCore.csproj [hunk 1]
```diff
@@ -15,12 +15,12 @@
     <GenerateAssemblyCompanyAttribute>false</GenerateAssemblyCompanyAttribute>
     <GenerateAssemblyProductAttribute>false</GenerateAssemblyProductAttribute>
     <Description>Workflow Core is a light weight workflow engine targeting .NET Standard.</Description>
-    <Version>3.3.5</Version>
-    <AssemblyVersion>3.3.5.0</AssemblyVersion>
-    <FileVersion>3.3.5.0</FileVersion>
+    <Version>3.3.6</Version>
```
Comment (danielgerlag) [inline src/WorkflowCore/WorkflowCore.csproj:18]: I think we need to bump the version of the MongoDb package instead

File: src/providers/WorkflowCore.Persistence.MongoDB/README.md [hunk 2]
```diff
@@ -17,3 +17,35 @@ Use the .UseMongoDB extension method when building your service provider.
 ```C#
 services.AddWorkflow(x => x.UseMongoDB(@"mongodb://localhost:27017", "workflow"));
 ```
+
+### State object serialization
```
Comment (danielgerlag) [inline src/providers/WorkflowCore.Persistence.MongoDB/README.md:21]: :+1:

File: src/WorkflowCore/WorkflowCore.csproj (hunk 1 above)
Comment (MrZoidberg) [reply src/WorkflowCore/WorkflowCore.csproj]: @danielgerlag fixed

**Pull request merged**